demos/bubble_shear.py

```diff
@@ -0,0 +1,129 @@
+    R = FunctionSpace(mesh, "R", 0)
+
+    c_ = Function(Q).project(c0)  # project initial condition onto the current mesh
+    c = Function(Q)  # solution at current timestep
+
+    t = Function(R).assign(t_start)
+    u_expression = velocity_expression(mesh, t)  # velocity at t_start
+    u_ = Function(V).interpolate(u_expression)
+    u = Function(V)  # velocity field at current timestep
+
+    # SUPG stabilisation
+    D = Function(R).assign(0.1)  # diffusivity coefficient
+    h = CellSize(mesh)  # mesh cell size
+    U = sqrt(dot(u, u))  # velocity magnitude
+    tau = 0.5 * h / U
+    tau = min_value(tau, U * h / (6 * D))
+
+    # Apply SUPG stabilisation to the test function
+    phi = TestFunction(Q)
+    phi += tau * dot(u, grad(phi))
+
+    # Time-stepping parameters
+    dt = Function(R).assign(0.01)  # timestep size
+    theta = Function(R).assign(0.5)  # Crank-Nicolson implicitness
+
+    # Variational form of the advection equation
+    trial = TrialFunction(Q)
+    a = inner(trial, phi) * dx + dt * theta * inner(dot(u, grad(trial)), phi) * dx
+    L = inner(c_, phi) * dx - dt * (1 - theta) * inner(dot(u_, grad(c_)), phi) * dx
+
+    # Define variational problem
+    lvp = LinearVariationalProblem(a, L, c, bcs=DirichletBC(Q, 0.0, "on_boundary"))
+    lvs = LinearVariationalSolver(lvp)
+
+    # Integrate from t_start to t_end
+    t.assign(t + dt)
+    while float(t) < t_end + 0.5 * float(dt):
+        # Update the background velocity field at the current timestep
+        u.interpolate(u_expression)
+
+        # Solve the advection equation
+        lvs.solve()
+
+        yield c
+
+        # Update the solution at the previous timestep
+        c_.assign(c)
+        u_.assign(u)
+        t.assign(t + dt)
+
+    return c
+
+
+def monitor(mesh):
+    # print('monitor mesh:', mesh)
+    alpha = Constant(50.0)
+
+    # _c = Function(FunctionSpace(mesh, "CG", 1)).interpolate(c_mov)
+
+    # P1_ten = TensorFunctionSpace(mesh, "CG", 1)
+    # H = RiemannianMetric(P1_ten)
+    # try:
+    #     H.compute_hessian(c_mov)
+    # except NameError:
+    #     H.compute_hessian(get_initial_condition(mesh))
+    H_interp = RiemannianMetric(TensorFunctionSpace(mesh, "CG", 1))
+    try:
+        H_interp.interpolate(H)
+    except NameError:
+        H_interp.interpolate(Constant(((1.0, 0.0), (0.0, 1.0))))
+
+    V = FunctionSpace(mesh, "CG", 1)
+    Hnorm = Function(V, name="Hnorm")
+    # Hnorm.interpolate(sqrt(inner(H, H)))
+    Hnorm.interpolate(sqrt(inner(H_interp, H_interp)))
+    Hnorm_max = Hnorm.dat.data.max()
+    m = 1 + alpha * Hnorm / Hnorm_max
+    return m
+
+
+simulation_end_time = T / 20.0  # 2.0
+mesh = UnitSquareMesh(64, 64)
+mover = MongeAmpereMover(mesh, monitor, method="quasi_newton", rtol=1e-8)
+c_mov = get_initial_condition(mover.mesh)
+movement_simulation = run_simulation(mover.mesh, 0.0, simulation_end_time, c_mov)
+t = 0
+H = RiemannianMetric(TensorFunctionSpace(mover.mesh, "CG", 1))
```

Comment:
Would this not fix your issue? Ensure `H` exists before it's needed.
```suggestion
H = RiemannianMetric(TensorFunctionSpace(mover.mesh, "CG", 1))
mover = MongeAmpereMover(mesh, monitor, method="quasi_newton", rtol=1e-8)
c_mov = get_initial_condition(mover.mesh)
movement_simulation = run_simulation(mover.mesh, 0.0, simulation_end_time, c_mov)
t = 0
```

Reply:
Thanks! Although I left this in more as a reminder to ask you whether we really need to call the `monitor` function during initialisation, since it gets called again during `mover.move()`. Seems like it could be optimised.

Reply:
Oh and maybe not, since `RiemannianMetric` (and the solution) needs to be defined on `mover.mesh` - I think?